### PR TITLE
Topic/embedded book

### PR DESCRIPTION
**Pull request by Robert-Tice**
PR for the embedded book

### CODE REVIEW

File: courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+      }
+
+      return 0;
+   }
+
+.. admonition:: The shortcuts of incrementing and decrementing
+
+   You may have noticed that Ada does not have something similar to the :c:`a++` or :c:`a--` operators. Instead you must use the full assignment :ada:`A := A + 1` or :ada:`A := A - 1`.
+
+**Fun Fact** about the C language assignment operator :c:`=`: Did you know that an assignment in C can be used in an expression? Let's look at an example:
```
Comment (gusthoff) [inline courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst]: I assume most C developers know this, so I'd rephrase this as a statement.

File: courses/Ada_For_The_C_Embedded_Developer/chapters/01_Introduction.rst [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+   #define DEGREES_MAX           (360)
+   typedef unsigned int degrees;
+
+   #define MOD_DEGREES(x)   (x % DEGREES_MAX)
+
+   degrees add_angles(degrees* list, int length)
+   {
+     degrees sum = 0;
+     for(int i = 0; i < length; ++i) {
```
Comment (gusthoff) [inline courses/Ada_For_The_C_Embedded_Developer/chapters/01_Introduction.rst]: Minor comment: I'd prefer a coding style where you use a whitespace after C-keywords (e.g.: ``for (...)``, ``if (...)``, etc).

File: courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+
+      if A = B then
+         Put_Line ("Yes");
+      else
+         Put_Line ("No");
+      end if;
+
+   end Main;
+
+If you try to run this Ada example you will get a compilation error. This is because the compiler is telling you that you cannot compare variables of two different types. We would need to explicitly cast one side to make the comparison against two variables of the same type. By enforcing the explicit cast we can't accidentally end up in a situation where we assume something will happen implicitly when, in fact, our assumption is incorrect.
```
Comment (gusthoff) [inline courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst]: At this point, some C-developers argue that a good C developer always pays attention and doesn't make this kind of mistakes. We should therefore emphasize that, although it's easy to spot the error in simple examples like this one, it becomes more complicated to find those bugs in larger code bases.

File: courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst (hunk 1 above)
Comment (Robert-Tice) [reply courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst]: I disagree that most C developers know this. There are CWE's about this:
https://cwe.mitre.org/data/definitions/481.html

and MISRA has an explicit rule about this. You're welcome to rephrase it into a statement though.

File: courses/Ada_For_The_C_Embedded_Developer/chapters/01_Introduction.rst [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      for I in List'Range loop
+         List (I) := Degrees (Integer'Value (Argument (I)));
+      end loop;
+
+      Put_Line ("Sum:" & Add_Angles (List)'Img);
+   end Sum_Angles;
+
+Here we have a piece of code in C and in Ada that takes some numbers from the command line and stores them in an array. We then sum all of the values in the array and print the result. The tricky part here is that we are working with values that model an angle in degrees. We know that angles are modular types, meaning that angles greater than 360° can also be represented as :ada:`Angle mod 360`. So if we have an angle of 400°, this is equivalent to 40°. In order to model this behavior in C we had to create the :c:`MOD_DEGREES` macro, which performs the modulus operation. As we read values from the command line, we convert them to integers and perform the modulus before storing them into the array. We then call add_angles which returns the sum of the values in the array. Can you spot the problem with the C code?
+
+Try running the Ada and C examples using the input sequence :ada:`340 2 50 70`. What does the C program output? What does the Ada program output? Why are they different?
```
[the remark this answers is not in the document]
Comment (Robert-Tice) [reply courses/Ada_For_The_C_Embedded_Developer/chapters/01_Introduction.rst]: Thats fine with me. I think that would make the text flow better.

File: courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst [hunk 5]
```diff
@@ -0,0 +1,802 @@
...
+
+The GNAT Toolchain
+-------------------
+
+The toolchain used throughout this book is called GNAT, which is a suite of tools with a compiler based on the GCC environment. It can be obtained from AdaCore, either as part of a commercial contract with `GNAT Pro <https://www.adacore.com/gnatpro)>`_ or at no charge with the `GNAT Community edition <https://www.adacore.com/community>`_. The information on this book will be relevant no matter which edition you’re using. Most examples will be runnable on the native Linux or Windows version for convenience. Some will only be relevant in the context of a cross toolchain, in which case we’ll be using the embedded ARM bare metal toolchain.
```
[the remark this answers is not in the document]
Comment (Robert-Tice) [reply courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst]: We might want to actually remove this section completely since we aren't asking the user to run anything on their system. I don't think its necessary to mention the FSF since we wouldn't be testing the code snippets on the FSF.

File: courses/Ada_For_The_C_Embedded_Developer/chapters/01_Introduction.rst (hunk 2 above)
Comment (Robert-Tice) [reply courses/Ada_For_The_C_Embedded_Developer/chapters/01_Introduction.rst]: Sure thats fine with me.
